stdlib/iarray.mli

```diff
@@ -0,0 +1,297 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*             Xavier Leroy, projet Cristal, INRIA Rocquencourt           *)
+(*               Antal Spector-Zabusky, Jane Street, New York             *)
+(*                                                                        *)
+(*   Copyright 1996 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*   Copyright 2023 Jane Street Group LLC                                 *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+open! Stdlib
+
+(* NOTE:
+   If this file is iarrayLabels.mli, run tools/sync_stdlib_docs after editing it
+   to generate iarray.mli.
+
+   If this file is iarray.mli, do not edit it directly -- edit
+   iarrayLabels.mli instead.
+ *)
+
+(** Operations on immutable arrays.  This module mirrors the API of [Array], but
+    omits functions that assume mutability; in addition to obviously mutating
+    functions, it omits [copy] along with the functions [make], [create_float],
+    and [make_matrix] that produce all-constant arrays.  The exception is the
+    sorting functions, which are given a copying API to replace the in-place
+    one. *)
+
+type +'a t = 'a iarray
+(** An alias for the type of immutable arrays. *)
+
+external length : 'a iarray -> int = "%array_length"
+(** Return the length (number of elements) of the given immutable array. *)
+
+external get : 'a iarray -> int -> 'a = "%array_safe_get"
+(** [get a n] returns the element number [n] of immutable array [a].
+   The first element has number 0.
+   The last element has number [length a - 1].
+   You can also write [a.:(n)] instead of [get a n].
+
+   @raise Invalid_argument
+   if [n] is outside the range 0 to [(length a - 1)]. *)
+
+external ( .:() ) : 'a iarray -> int -> 'a = "%array_safe_get"
```

Comment:
I find the use of `:` visually disturbing on reading (I associate it more to the type level language). Did you consider perhaps using `@` or `%` ?

Reply:
Normally I wouldn’t have a particular preference on this point. In this specific case, I personally still hope that we could add syntax for immutable arrays later (in a future PR, if usage shows that it would be useful). In this hypothetical case, and assuming that this hypothetical future syntax would look like the `[: ... :]` initially proposed here, `.:()` as an access operator would look more consistent than `.@()` or `.%()`. If `.:()` evokes type-related stuff too much, we could go for `.::()` instead.

For the record, I don’t find `.:()` particularly type-evocating.

Reply:
> we could go for .::() instead.

That evokes lists… and the point of an indexing operator is that it should be short, at that point you can use `get`… 

> For the record, I don’t find `.:()` particularly type-evocating.

In the language it is only ever used at the type language level (except if you use custom indexing operators). It is a strong signifier.

Perhaps leave the operator out of proposal ? FWIW there is also no syntax for indexing `bytes` and I don't find myself utterly missing that.

Reply:
Since the indexing operator can raise an exception, what about `.!( )` ?

Reply:
I'd rather keep `!` for mutable stuff :–)

Reply:
Beyond types, `:` also appears in label syntax, in both types and terms.

In a better world, we could use `[| … |]` for immutable arrays and `[! … !]` for mutable ones, but we’re not in that world.

Reply:
Personally, I'd prefer not to add an indexing operator at this point, and to instead investigate extending array disambiguation to support indexing (which has already been implemented once, in #616).  It'd be better to have a uniform scoped/disambiguated syntax for indexing rather than lots of individual operators with funny names.